### PR TITLE
Fixes SR-642: Code completion does not instantiate generic arguments of a type wrapped in an optional.

### PR DESCRIPTION
**Pull request by ahoppen**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
Could you add a test?

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -1708,7 +1708,7 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
       Type ContextTy = VD->getDeclContext()->getDeclaredTypeOfContext();
       if (ContextTy) {
         Type MaybeNominalType = ExprType->getRValueInstanceType();
-        if (ContextTy->getAnyNominal() == MaybeNominalType->getAnyNominal() &&
+        if (ContextTy->getDesugaredType()->getAnyNominal() == MaybeNominalType->getDesugaredType()->getAnyNominal() &&
```
Comment (gribozavr) [inline lib/IDE/CodeCompletion.cpp]: 80 columns.

**Comment by gribozavr** [comment]
... and it would be best if the test didn't use Set, but used a mock type instead.  `Set` can change in future.

**Comment by ahoppen** [comment]
I added a test case and added a line break so that the lines are only 80-ish columns. Thanks for the hint with the mock type, wouldn’t have thought about that

File: test/IDE/complete_generic_optional.swift [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+struct Foo<T> {
+	func myFunction(foobar: T) {
+	}
+}
+
+// SR-642 Code completion does not instantiate generic arguments of a type wrapped in an optional
+let x: Foo<Bar>? = Foo<Bar>()
+x.#^FOO_OPTIONAL_1^#
+// FOO_OPTIONAL_1: Begin completions, 4 items
+// FOO_OPTIONAL_1-DAG: ecl[InstanceMethod]/CurrNominal/Erase[1]: ?.myFunction({#(foobar): Bar#})[#Void#]; name=myFunction(foobar: Bar)
```
Comment (NachoSoto) [inline test/IDE/complete_generic_optional.swift]: This is awesome :+1:

Comment (gribozavr) [inline test/IDE/complete_generic_optional.swift]: *D*ecl?

File: test/IDE/complete_generic_optional.swift [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=FOO_OPTIONAL_1 | FileCheck %s -check-prefix=FOO_OPTIONAL_1
+
+struct Bar {
+}
+
+struct Foo<T> {
+	func myFunction(foobar: T) {
+	}
```
Comment (gribozavr) [inline test/IDE/complete_generic_optional.swift]: Please use two spaces for indentation.

**Comment by ahoppen** [comment]
Sorry, don't know how these changes happened between my tests and and the commit. Hope everything is fine now ;-)

**Comment by tkremenek** [comment]
@swift-ci Please test

**Pull request merged**